kvm/kvm_test.go

```diff
@@ -407,3 +407,421 @@ func TestIoctlStringer(t *testing.T) {
+
+	devKVM, err := os.OpenFile("/dev/kvm", os.O_RDWR, 0o644)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer devKVM.Close()
+
+	vmFd, err := kvm.CreateVM(devKVM.Fd())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cd := &kvm.ClockData{}
+
+	if err := kvm.GetClock(vmFd, cd); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := kvm.SetClock(vmFd, cd); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestCreateDev(t *testing.T) {
+	if os.Getuid() != 0 {
+		t.Skipf("Skipping test since we are not root")
+	}
+
+	t.Parallel()
+
+	devKVM, err := os.OpenFile("/dev/kvm", os.O_RDWR, 0o644)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer devKVM.Close()
+
+	vmFd, err := kvm.CreateVM(devKVM.Fd())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	dev := &kvm.Device{
+		Type:  uint32(kvm.DevVFIO),
+		Fd:    0,
+		Flags: 1,
+	}
+
+	for i := 0; i <= int(kvm.DevMAX); i++ {
+		if err = kvm.CreateDev(vmFd, dev); err != nil {
+			if !errors.Is(err, syscall.ENODEV) {
+				t.Fatal(err)
+			}
+		}
+	}
+}
+
+func TestInjectInterrpt(t *testing.T) {
+	if os.Getuid() != 0 {
+		t.Skipf("Skipping test since we are not root")
+	}
+
+	t.Parallel()
+
+	devKVM, err := os.OpenFile("/dev/kvm", os.O_RDWR, 0o644)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer devKVM.Close()
+
+	vmFd, err := kvm.CreateVM(devKVM.Fd())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	mem, err := syscall.Mmap(-1, 0, 0x1000, syscall.PROT_READ|syscall.PROT_WRITE, syscall.MAP_SHARED|syscall.MAP_ANONYMOUS)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if err = kvm.SetUserMemoryRegion(vmFd, &kvm.UserspaceMemoryRegion{
+		Slot:          0,
+		Flags:         0,
+		GuestPhysAddr: 0x1000,
+		MemorySize:    0x1000,
+		UserspaceAddr: uint64(uintptr(unsafe.Pointer(&mem[0]))),
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	vcpuFd, err := kvm.CreateVCPU(vmFd, 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Pass an invalid value, because the vm is empty and error out for every other error
+	if err := kvm.InjectInterrupt(vcpuFd, 0xFFF0); !errors.Is(err, syscall.EFAULT) {
+		t.Fatal(err)
+	}
+}
+
+func TestGetMSRIndexList(t *testing.T) {
+	if os.Getuid() != 0 {
+		t.Skipf("Skipping test since we are not root")
+	}
+
+	t.Parallel()
+
+	devKVM, err := os.OpenFile("/dev/kvm", os.O_RDWR, 0o644)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer devKVM.Close()
+
+	list := kvm.MSRList{}
+	list.NMSRs = 100
+
+	_ = kvm.GetMSRIndexList(devKVM.Fd(), &list)
+}
+
+func TestGetMSRFeatureIndexList(t *testing.T) {
+	if os.Getuid() != 0 {
+		t.Skipf("Skipping test since we are not root")
+	}
+
+	t.Parallel()
+
+	devKVM, err := os.OpenFile("/dev/kvm", os.O_RDWR, 0o644)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer devKVM.Close()
+
+	list := kvm.MSRList{}
+	list.NMSRs = 100
+
+	if err = kvm.GetMSRFeatureIndexList(devKVM.Fd(), &list); err != nil {
+		t.Fatal(err)
+	}
+
+	var entryFound bool
+
+	for _, msr := range list.Indicies {
+		if msr != 0 {
+			entryFound = true
+		}
+	}
+
+	if !entryFound {
+		t.Log("no entry has been found and no error occurred. That's odd")
```

Comment:
Same here, it would be better to skip this test only when KVM_CAP_GET_MSR_FEATURES is disabled.

Reply:
done